### PR TITLE
Improve temperature runaway and idle timeout

### PR DESCRIPTION
**Pull request by ellensp**
### Requirements

THERMAL_PROTECTION_HOTENDS
THERMAL_PROTECTION_BED
THERMAL_PROTECTION_CHAMBER

### Description

With these enabled if heater  > MAXTEMP  it gives a MSG_THERMAL_RUNAWAY  
This is not very useful. 
changed to MAXTEMP errors. 
Separated bed and chamber thermal protection that where incorrectly share same array element. 
Disable thermal protection if device doesn't exist.   

### Benefits

These 3 errors are more specific as to what is going on.
Chamber and bed thermal protection is independent.

### Related Issues

Was noticed by AnHardt  in https://github.com/MarlinFirmware/Marlin/issues/19335

### CODE REVIEW

**Comment by thinkyhead** [comment]
The Maxtemp and Mintemp errors are specifically about:
- a thermistor that is disconnected, so it is seeing a maximum (♾️) resistance value, and
- a thermistor that is shorted, so the hardware is sending the minimum (zero) resistance value.

Those errors should not be thrown for these conditions.

The existing errors pertain to a "thermal runaway", where the heater is not behaving like it should be. These tests check for a runaway by checking if the heater has gone outside the hotend's max temperature tolerance. The error could be due to a hardware fault, or maybe the PID tuning is really bad. A thermal runaway alerts the user to look for:
- a MOSFET/heater that is failed-on,
- a MOSFET/heater that is failed-off, or
- a thermistor that has fallen out (so the temperature doesn't rise when the heater is on).

**Comment by AnHardt** [comment]
@thinkyhead 

> The Maxtemp and Mintemp errors are specifically about:

a thermistor that is disconnected, so it is seeing a maximum (♾️) resistance value, and
a thermistor that is shorted, so the hardware is sending the minimum (zero) resistance value.

No.
The raw limits variables used to be initialized at the wrong end and then always have ben adjusted to the configured max and mins from the configs with a kind of this loops.
```
  #if HAS_HEATED_BED
    #ifdef BED_MINTEMP
      while (analog_to_celsius_bed(mintemp_raw_BED) < BED_MINTEMP) mintemp_raw_BED += TEMPDIR(BED) * (OVERSAMPLENR);
    #endif
    #ifdef BED_MAXTEMP
      while (analog_to_celsius_bed(maxtemp_raw_BED) > BED_MAXTEMP) maxtemp_raw_BED -= TEMPDIR(BED) * (OVERSAMPLENR);
    #endif
  #endif // HAS_HEATED_BED
```
Max and min errors always have been thrown above and below max and min temperatures not only at the ADCs end values.

It's your decision if we want to use the one or the other definition, but currently we have both definitions at the same time. And i guess it's not by will, but by incompetence.

> a MOSFET/heater that is failed-on,

I'd be glad if my analyse would come to this result - but it does not! **Please** have a look on (https://github.com/MarlinFirmware/Marlin/issues/19335#issuecomment-690689900). I may be completely wrong but i don't think so.

**Comment by GMagician** [comment]
@ellensp may you also fix the temperature array size/index error? it should be HOTENDS + 1 + 1 (where the double +1 depends on BED or CHAMBER are checked)

**Comment by GMagician** [comment]
I would like to check how to change all indexing using header_t enum, but I wouldn't do a parallel work with other devs.
In such case this PR may be considered mergeable. Is there someone looking for such change?

**Comment by GMagician** [comment]
@ellensp & @thinkyhead I'm thinking about using `heater_ind_t`, properly filled, to convert to index into arrays. I see that such typedef is used in a lot of places. and in more than one is used `<0` to identify bed, but, when chamber may be used there, this may lead to issues. I think it will be a complex change (find everywhere such typedef and replace hotend index to properly value, a +/- H_E0 must be used but if we keep it 0 no code overhead should be added)

**Comment by ellensp** [comment]
After making tr_target_temperature dynamically sized, with no hotend, heated bed or chamber code was still trying to setup   thermal_runaway_protection. This has now been addressed.

**Comment by GMagician** [comment]
I know it has been adressd but I was thinking to a "more" robust solution to use on all code, not just here.

**Comment by ellensp** [comment]
@GMagician If you want to do that, go for it. I'm just trying to squash a few bug without making a mess.

**Comment by GMagician** [comment]
> I'm just trying to squash a few bug without making a mess

I agree with you, mine was just an idea trying to avoid something similar in other parts of the code, I was just asking opinions, It may be a long job to discover all parts and I don't want to lose time without any good reason

**Comment by AnHardt** [comment]
So far, thanks for that.

Remaining things to think about:

Do you want the max/min-temp tests to be redundant - once at the raw ADC values or at the incoming MAX???? values **and** here where you have changed now?
Do you really want to catch a "stuck ON heater" only by detecting maxtemp or do you want to catch it in `thermal_runaway_protection` what the name suggests [but it wasn't meant for](https://github.com/MarlinFirmware/Marlin/commit/43c298a7a9674bade1395a75fe54667cdb22fce4)? Currently it does only detect broken (stuck OFF) heaters and fallen out temperature sensors. How about just checking against a window around the target temperature?
Isn't what "watching heaters" does, testing if a heater **does** heat, only meaningful in the `TRFirstHeating`-state. Isn't the "start watching heater" called exactly at the transition to `TRFirstHeating` state? Could that save some code-size or runtime?

File: Marlin/src/module/temperature.cpp [hunk 1]
```diff
@@ -1962,11 +1964,12 @@ void Temperature::init() {
...
     //*/
 
-    #if BOTH(HAS_HEATED_BED,HAS_HEATED_CHAMBER)
-      const int heater_index = heater_id >= 0 ? heater_id : heater_id == H_CHAMBER ? HOTENDS+1 : HOTENDS;
-    #else
-      const int heater_index = heater_id >= 0 ? heater_id : HOTENDS;
-    #endif
+    const uint8_t heater_index = (heater_id >= 0 ? heater_id :
+      #if BOTH(HAS_HEATED_BED, HAS_HEATED_CHAMBER)
+        heater_id == H_CHAMBER ? HOTENDS + 1 :
```
Comment (GMagician) [inline Marlin/src/module/temperature.cpp:1969]: Isn't this a bad index due to H_CHAMBER == -1?

**Pull request merged**